File: James/221227_study/프로_12978_배달.java
import java.util.*;

class Solution {
    static class Node {
        int b, cost;
        
        Node(int b, int cost) {
            this.b = b;
            this.cost = cost;
        }
    }
    
    static ArrayList<ArrayList<Node>> graph = new ArrayList<>();
    static int[] distance;
    
    public static void dijkstra(int start) {
        PriorityQueue<Node> pq = new PriorityQueue<>((a, b) -> Integer.compare(a.cost, b.cost)); 
        distance[start] = 0;
        pq.add(new Node(start, 0));
        
        while(!pq.isEmpty()) {
            Node data = pq.poll();
            int now = data.b;
            int dist = data.cost;
            
            if(distance[now] < dist)
                continue;
            
            for(Node n : graph.get(now)) {
                int node = n.b;
                int weight = n.cost;
                int cost = dist + weight;
                
                if(distance[node] > cost) {
                    distance[node] = cost;
                    pq.add(new Node(node, cost));
                }
            }
        }
    }
    
    public int solution(int N, int[][] road, int K) {
        int answer = 0;
        distance = new int[N + 1];
        
        for(int i = 0; i < N + 1; i++) {
            graph.add(new ArrayList<>());
            distance[i] = Integer.MAX_VALUE;
        }
        
        for(int i = 0; i < road.length; i++) {
            int a = road[i][0];
            int b = road[i][1];
            int c = road[i][2];
            graph.get(a).add(new Node(b, c));
            graph.get(b).add(new Node(a, c));
        }
        
        dijkstra(1);
        
        for(int i = 0; i < distance.length; i++) {
            if(distance[i] <= K)
                answer++;
        }
        
        return answer;
    }
}

File: sammy/221227_study/프로_12978_배달.py
import heapq

def solution(N, road, K):
    answer = 0

    INF = int(1e9) # 무한을 의미하는 값으로 10억을 설정
    
    # 각 노드에 연결된 노드에 대한 정보를 담는 리스트 생성
    graph = [[] for i in range(N+1)]
    
    # 최단 거리 테이블을 모두 무한으로 초기화
    distance = [INF] * (N+1)
    
    for i in road:
        graph[i[0]].append((i[1],i[2]))
        graph[i[1]].append((i[0],i[2]))
        
    q = []
    
    # 시작 노드로 가기 위한 최단 경로는 0으로 설정하여, 큐에 삽입
    heapq.heappush(q, (0, 1))
    distance[1] = 0
    
    while q: # 큐가 비어있지 않다면
        
        # 최단 거리가 가장 짧은 노드에 대한 정보 꺼내기
        dist, now = heapq.heappop(q)
        
        # 현재 노드가 이미 처리된 적이 있는 노드라면 무시
        if distance[now] < dist:
            continue
        
        # 현재 노드와 연결된 다른 인접한 노드들을 확인
        for i in graph[now]:
            cost = dist + i[1]
            
            # 현재 노드를 거쳐서, 다른 노드로 이동하는 거리가 더 짧은 경우
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))

    for check in distance: # distance가 K 이하인 값 개수 구하기
        if check<=K:
            answer+=1
    return answer

print(solution(5,[[1,2,1],[2,3,3],[5,2,2],[1,4,2],[5,3,1],[5,4,2]],3))
print(solution(6,[[1,2,1],[1,3,2],[2,3,2],[3,4,3],[3,5,2],[3,5,3],[5,6,1]],4))